```{r setup}
knitr::opts_chunk$set(
	error = TRUE,
	message = TRUE
)

mycolors = c("#69d2e7", "#a7dbd8", "#e0e4cc", "#f38630", "#fa6900", "#fc9d9a", "#f9cdad", "#c8c8a9", "#83af9b", "#ecd078", "#d95b43", "#542437", "#53777a")
```

```{r}
set.seed(123)

library(phyloseq)
library(microeco)
library(file2meco)
library(tidyverse)
library(microbiome)
library(knitr)
library(DT)
library(RColorBrewer)
library(microViz)
library(ggsankey)
library(ComplexHeatmap)
```
## Load phyloseq object
```{r}
physeq <- readRDS("physeq_16s_160-142.rds")
```
```{r}
metadata <- read.table("sampledata_16s_meta.txt", header = T, row.names = 1, sep = "\t")
stats_dada2 = readRDS("stats_dada2_16s.rds")
metadata = merge(metadata, stats_dada2, by=0, all.x=TRUE)
rownames(metadata) = metadata[,1]
sample_data(physeq) <- metadata
```
## Transform phyloseq object to microtable class
```{r}
meco <- phyloseq2meco(physeq)
meco$sample_table = filter(meco$sample_table, Season != "Autumn")
meco$filter_pollution(taxa = c("mitochondria", "chloroplast"))
meco$tax_table = filter(meco$tax_table, Kingdom != "k__Eukaryota", Kingdom != "k__")
meco$tidy_dataset()
```
# Paired-end reads were obtained (sum)
```{r}
sum(meco$sample_table$input)
```
# Filtered sequances (sum)
```{r}
sum(meco$otu_table)
```
# Filtered sequances (mean)
```{r}
mean(colSums(meco$otu_table))
```
# Filtered sequances (median)
```{r}
median(colSums(meco$otu_table))
```

<!-- ```{r}
sampledata = merge(meco$sample_table, as.data.frame(colSums(meco$otu_table)), by=0, all.x=TRUE)
write.table(sampledata, "sampledata_16_mod.txt", quote=F, sep = "\t")
``` -->

## Fix tax table
```{r}
tax_meta = meco$tax_table

tax_rows = rownames(tax_meta)
t = apply(tax_meta, 2, function(x) str_replace_all(x, ".__$", "uncultured"))
rownames(t) = tax_rows
tax_meta = t

# Swap values ​​from Swap_unidentified in tax_meta to previous taxon level
Swap_unidentified = c("uncultured", "unidentified", "metagenome", "bacteriap25", "Unknown", "Incertae_Sedis")
for(Swap in Swap_unidentified) {
    for(j in 1:7) {
        for (i in 1:nrow(tax_meta)) {
            if(grepl(Swap, tax_meta[i,j])) {
                tax_meta[i,j] = tax_meta[i,j-1] }
                else { tax_meta[i,j] = tax_meta[i,j]
            }       
        }
    }
}
tax_meta = as.data.frame(tax_meta)

# Mark unclassified taxa at the genus level as "un."
tax_meta$Phylum = str_replace_all(tax_meta$Phylum, "k__.*", paste(tax_meta$Phylum, "un."))
tax_meta$Class = str_replace_all(tax_meta$Class, "k__.*", paste(tax_meta$Class, "un."))
tax_meta$Class = str_replace_all(tax_meta$Class, "p__.*", paste(tax_meta$Class, "un."))
tax_meta$Order = str_replace_all(tax_meta$Order, "k__.*", paste(tax_meta$Order, "un."))
tax_meta$Order = str_replace_all(tax_meta$Order, "p__.*", paste(tax_meta$Order, "un."))
tax_meta$Order = str_replace_all(tax_meta$Order, "c__.*", paste(tax_meta$Order, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "k__.*", paste(tax_meta$Family, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "p__.*", paste(tax_meta$Family, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "c__.*", paste(tax_meta$Family, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "o__.*", paste(tax_meta$Family, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "k__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "p__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "c__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "o__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "f__.*", paste(tax_meta$Genus, "un."))

tax_meta$Class_Genus = paste0(tax_meta$Class, "; ", tax_meta$Genus)
tax_meta = tax_meta[,c("Kingdom", "Phylum", "Class", "Order", "Family", "Class_Genus", "Genus", "Species")]

Species = paste("[ASV", c(1:length(rownames(meco$tax_table))), "]", sep = "")
Species = paste(Species, tax_meta$Species, sep = " ")
meco$tax_table = as.data.frame(tax_meta)
meco$tax_table$Species = Species

```

## Rarefaction
```{r}
meco_rf <- clone(meco)
meco_rf$rarefy_samples(sample.size = 10000)
meco_rf$tidy_dataset()
```

# Alpha diversity
```{r}
tmp <- clone(meco_rf)
```
## Shannon (Organ.material)
```{r fig.width=5, fig.height=6}
t1 <- trans_alpha$new(dataset = tmp, group = "Organ.material")
t1$cal_diff(method = "anova")
t1$plot_alpha(measure = "Shannon") +
	scale_y_continuous(breaks = c(0,1,2,3,4,5,6), limits = c(2,6))
ggsave("../fig 1a. Shannon (Organ.material) (16S).png", width=3, height=4, dpi=400)
```
## Shannon (Season)
```{r fig.width=5, fig.height=6}
t1 <- trans_alpha$new(dataset = tmp, group = "Season")
t1$cal_diff(method = "anova")
t1$plot_alpha(measure = "Shannon") +
	scale_y_continuous(breaks = c(0,1,2,3,4,5,6), limits = c(2,6))
ggsave("../fig 1b. Shannon (Season) (16S).png", width=3, height=4, dpi=400)
```
## Shannon (Plant)
```{r fig.width=5, fig.height=6}
t1 <- trans_alpha$new(dataset = tmp, group = "Plant")
t1$cal_diff(method = "anova")
t1$plot_alpha(measure = "Shannon") +
	scale_y_continuous(breaks = c(0,1,2,3,4,5,6), limits = c(2,6))
ggsave("../fig 1c. Shannon (Plant) (16S).png", width=3, height=4, dpi=400)
```

# Beta diversity
```{r }
tmp <- clone(meco_rf)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
tmp$tidy_dataset()
tmp$cal_betadiv()
```
## Bray-curtis (Organ.material)
```{r fig.width=6, fig.height=6}
t1 <- trans_beta$new(dataset = tmp, group = "Organ.material", measure = "bray")
t1$cal_ordination(method = "PCoA")
t1$plot_ordination(plot_color = "Organ.material", plot_shape = "Plant", plot_type = c("point", "ellipse")) + 
	guides(
		color = guide_legend(order = 1),
		fill = guide_legend(order = 1),
		shape = guide_legend(order = 2)
	) + 
	labs(
		color = "Organ material",
		fill = "Organ material"
	)
ggsave("../fig 1d. Bray-curtis (Organ.material) (16S).png", width=4, height=4, dpi=400)
```
## PERMANOVA (Organ.material)
```{r }
t1$cal_manova(group = "Organ.material", manova_all = F)
kable(t1$res_manova)
```
## Bray-curtis (Season)
```{r fig.width=6, fig.height=6}
t1 <- trans_beta$new(dataset = tmp, group = "Season", measure = "bray")
t1$cal_ordination(method = "PCoA")
t1$plot_ordination(plot_color = "Season", plot_shape = "Plant", plot_type = c("point", "ellipse")) + 
	guides(
		color = guide_legend(order = 1),
		fill = guide_legend(order = 1),
		shape = guide_legend(order = 2)
	) + 
	labs(
		color = "Season",
		fill = "Season"
	)
ggsave("../fig 1e. Bray-curtis (Season) (16S).png", width=4, height=4, dpi=400)
```
## PERMANOVA (Season)
```{r }
t1$cal_manova(group = "Season", manova_all = F)
kable(t1$res_manova)
```
## Bray-curtis (Plant)
```{r fig.width=6, fig.height=6}
t1 <- trans_beta$new(dataset = tmp, group = "Plant", measure = "bray")
t1$cal_ordination(method = "PCoA")
t1$plot_ordination(plot_color = "Plant", plot_shape = "Organ.material", plot_type = c("point", "ellipse")) + 
	guides(
		color = guide_legend(order = 1),
		fill = guide_legend(order = 1),
		shape = guide_legend(order = 2)
	) + 
	labs(
		color = "Plant",
		fill = "Plant"
	)
ggsave("../fig 1f. Bray-curtis (Plant) (16S).png", width=4, height=4, dpi=400)
```
## PERMANOVA (Plant)
```{r }
t1$cal_manova(group = "Plant", manova_all = F)
kable(t1$res_manova)
```


# Composition
```{r}
tmp <- clone(meco_rf)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
tmp$tidy_dataset()

```
## Organ material bar plot (Phylum)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Organ.material")
t1 <- trans_abund$new(dataset = bar, taxrank = "Class", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
t1$plot_bar(others_color = "grey70", xtext_keep = TRUE, legend_text_italic = FALSE,
color_values = mycolors,
order_x = c("Branch", "Needles", "Shavings")) +
geom_text(aes(x = Sample, y = Abundance, 
		label = ifelse(Abundance > 2, scales::percent(Abundance, scale = 1, accuracy = 1), ""), 
		group = Taxonomy), 
		inherit.aes = FALSE, position = position_stack(vjust = 0.5), col = "black")
ggsave("../fig 2a. Organ material bar plot (Class) (16S).png", width=5, height=6, dpi=400)
```

## Organ material venn diagramm
```{r fig.width=5, fig.height=5}
ven = clone(bar)

tax = as.data.frame(ven$tax_table[,"Species"])
rownames(tax) = rownames(ven$tax_table)
colnames(tax) = "Species"
ven$tax_table = tax
ven = ven$merge_taxa("Species")

rownames(ven$tax_table) = ven$tax_table$Species
rownames(ven$otu_table) = ven$tax_table$Species
t1 <- trans_venn$new(ven, ratio = "seqratio")
t1$plot_venn()
ggsave("../fig 2d. Organ material venn diagramm (ASV) (16S).png", width=5, height=5, dpi=400)
```
```{r}
datatable(t1$data_details, options = list(scrollX = TRUE, scrollY = TRUE))
write.table(t1$data_details, sep = "\t", quote = FALSE, file="Organ material venn diagramm (ASV) (16S).txt")
```


## Season bar plot (Class)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Season")
t1 <- trans_abund$new(dataset = bar, taxrank = "Class", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
t1$plot_bar(others_color = "grey70", xtext_keep = TRUE, legend_text_italic = FALSE,
color_values = mycolors,
order_x = c("Summer", "Winter")) +
geom_text(aes(x = Sample, y = Abundance, 
		label = ifelse(Abundance > 2, scales::percent(Abundance, scale = 1, accuracy = 1), ""), 
		group = Taxonomy), 
		inherit.aes = FALSE, position = position_stack(vjust = 0.5), col = "black")
ggsave("../fig 2b. Season bar plot (Class) (16S).png", width=5, height=6, dpi=400)
```

## Season venn diagramm
```{r fig.width=5, fig.height=5}
ven = clone(bar)

tax = as.data.frame(ven$tax_table[,"Species"])
rownames(tax) = rownames(ven$tax_table)
colnames(tax) = "Species"
ven$tax_table = tax
ven = ven$merge_taxa("Species")

rownames(ven$tax_table) = ven$tax_table$Species
rownames(ven$otu_table) = ven$tax_table$Species
t1 <- trans_venn$new(ven, ratio = "seqratio")
t1$plot_venn()
ggsave("../fig 2e. Season venn diagramm (ASV) (16S).png", width=5, height=5, dpi=400)
```
```{r}
datatable(t1$data_details, options = list(scrollX = TRUE, scrollY = TRUE))
write.table(t1$data_details, sep = "\t", quote = FALSE, file="Season venn diagramm (ASV) (16S).txt")
```
## Plant bar plot (Class)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Plant")
t1 <- trans_abund$new(dataset = bar, taxrank = "Class", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
t1$plot_bar(others_color = "grey70", xtext_keep = TRUE, legend_text_italic = FALSE,
color_values = mycolors,
order_x = c("Pj_1", "Pj_2")) +
geom_text(aes(x = Sample, y = Abundance, 
		label = ifelse(Abundance > 2, scales::percent(Abundance, scale = 1, accuracy = 1), ""), 
		group = Taxonomy), 
		inherit.aes = FALSE, position = position_stack(vjust = 0.5), col = "black")
ggsave("../fig 2c. Plant bar plot (Phylum) (16S).png", width=5, height=6, dpi=400)
```

## Plant venn diagramm
```{r fig.width=5, fig.height=5}
ven = clone(bar)

tax = as.data.frame(ven$tax_table[,"Species"])
rownames(tax) = rownames(ven$tax_table)
colnames(tax) = "Species"
ven$tax_table = tax
ven = ven$merge_taxa("Species")

rownames(ven$tax_table) = ven$tax_table$Species
rownames(ven$otu_table) = ven$tax_table$Species
t1 <- trans_venn$new(ven, ratio = "seqratio")
t1$plot_venn()
ggsave("../fig 2f. Plant venn diagramm (ASV) (16S).png", width=5, height=5, dpi=400)
```
```{r}
datatable(t1$data_details, options = list(scrollX = TRUE, scrollY = TRUE))
write.table(t1$data_details, sep = "\t", quote = FALSE, file="Plant venn diagramm (ASV) (16S).txt")
```

## ASVs summary
```{r}
length(rownames(tmp$tax_table))
```
```{r}
k = tmp$tax_table %>%
	group_by(Genus) %>%
	reframe(num = length(Genus)) %>%
	arrange(desc(num)) %>%
	mutate(group = paste0(Genus, " (", num, " ASVs)"))
```
```{r}
separate_wider_delim(k, cols = "Genus", names_sep = "_", delim  = "__") %>%
	group_by(Genus_1) %>%
	reframe(num = sum(num))
```
```{r}
tmp$tax_table %>%
	group_by(Genus) %>%
	reframe(num = length(Genus)) %>%
	arrange(desc(num)) %>%
	mutate(group = paste0(Genus, " (", num, " ASVs)"))
```
```{r}
tmp$tax_table %>%
	group_by(Genus) %>%
	reframe(num = length(Genus)) %>%
	arrange(desc(num)) %>%
	mutate(group = paste0(Genus, " (", num, " ASVs)")) %>%
	as.data.frame()
```

## Core taxa
```{r}
core = clone(tmp)
tax = as.data.frame(core$tax_table[,"Class_Genus"])
rownames(tax) = rownames(core$tax_table)
colnames(tax) = "Class_Genus"
core$tax_table = tax
core = core$merge_taxa("Class_Genus")

core <- meco2phyloseq(core)
pseq.rel <- microbiome::transform(core, "compositional")
taxa_names(pseq.rel) = tax_table(pseq.rel)[,"Class_Genus"]

# Core with compositionals:
prevalences <- seq(.05, 1, .05)
detections <- round(10^seq(log10(1e-2), log10(.2), length = 10), 3)
# detections <- c(0.005, round(10^seq(log10(1e-2), log10(.2), length = 10), 3))

# Also define gray color palette
gray <- gray(seq(0,1,length=5))

p1 <- plot_core(pseq.rel,
  plot.type = "heatmap",
  colours = rev(brewer.pal(5, "RdBu")),
  prevalences = prevalences,
  detections = detections, min.prevalence = .5) +
  xlab("Detection Threshold (Relative Abundance (%))")

p1 <- p1 + theme_bw() + ylab("Taxa")
p1
ggsave("../fig 3. core (16S).png", width=8, height=5, dpi=400)
```

## Differential abundance test (Organ.material)
```{r}
tmp <- clone(meco)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
tmp$tidy_dataset()
```
```{r }
difab = clone(tmp)
tax = as.data.frame(difab$tax_table[,"Class_Genus"])
rownames(tax) = rownames(difab$tax_table)
colnames(tax) = "Class_Genus"
difab$tax_table = tax
difab = difab$merge_taxa("Class_Genus")

t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Organ.material", fix_formula = "Organ.material", taxa_level = "Class_Genus", pairwise = TRUE)

res_convert <- data.frame()
tmp_res = t1$res_diff_raw$res_pair
for(i in seq_len(nrow(tmp_res))){
	taxon_data <- tmp_res[i, ]
	raw_colnames <- colnames(taxon_data)
	all_factors <- raw_colnames %>% .[grepl("lfc_", .)] %>% gsub("lfc_", "", .)
	res_convert <- rbind(res_convert, data.frame(Taxa = tmp_res[i, "taxon"], 
		Factors = all_factors, 
		lfc = taxon_data %>% .[, grepl("^lfc_", raw_colnames)] %>% unlist,
		se = taxon_data %>% .[, grepl("^se_", raw_colnames)] %>% unlist,
		W = taxon_data %>% .[, grepl("^W_", raw_colnames)] %>% unlist,
		p = taxon_data %>% .[, grepl("^p_", raw_colnames)] %>% unlist,
		P.adj = taxon_data %>% .[, grepl("^q_", raw_colnames)] %>% unlist,
		diff = taxon_data %>% .[, grepl("^diff_", raw_colnames)] %>% unlist,
		passed_ss = taxon_data %>% .[, grepl("^passed_ss_", raw_colnames)] %>% unlist
		)
	)
}

t1$res_diff = res_convert

res = as_tibble(t1$res_diff) %>%
	arrange(Taxa) %>%
	group_by(Taxa) %>%
	filter(sum(P.adj < 0.05) >= 2) %>%
	ungroup()
```

```{r }
diff = str_split(unique(filter(res, passed_ss == T & diff == T)$Taxa), "\\|", simplify = TRUE)

Needles = str_split(unique(filter(res, Factors == "Organ.materialNeedles" & lfc > 0 & passed_ss == T & diff == T | 
	Factors == "Organ.materialShavings_Organ.materialNeedles" & lfc < 0 & passed_ss == T & diff == T )$Taxa ), "\\|", simplify = TRUE)

Shavings = str_split(unique(filter(res, Factors == "Organ.materialShavings" & lfc > 0 & passed_ss == T & diff == T | 
	Factors == "Organ.materialShavings_Organ.materialNeedles" & lfc > 0 & passed_ss == T & diff == T )$Taxa ), "\\|", simplify = TRUE)

Branch = str_split(unique(filter(res, Factors == "Organ.materialShavings" & lfc < 0 & passed_ss == T & diff == T  | 
	Factors == "Organ.materialNeedles" & lfc < 0 & passed_ss == T & diff == T )$Taxa ), "\\|", simplify = TRUE)

Needles = as_tibble(Needles)
Needles$diff = "Needles"

Shavings = as_tibble(Shavings)
Shavings$diff = "Shavings"

Branch = as_tibble(Branch)
Branch$diff = "Branch"

all = rbind(Needles, Shavings, Branch) %>%
	arrange(V1) %>%
	group_by(V1) %>%
	nest(data = diff) %>%
	reframe(type = toString(unlist(data)))

diff_all = all$type
names(diff_all) = all$V1

psq <- meco2phyloseq(tmp)

psq = psq %>%
	merge_samples("Organ.material") %>%
	tax_transform("compositional", rank = "Class_Genus") %>%
	tax_select(diff, ranks_searched = "Class_Genus", strict_matches = T) 

diff_all = diff_all[taxa_names(psq)]

col_fun = distinct_palette(n = length(unique(all$type)), add = NA)
names(col_fun) =  unique(all$type)

png("../fig 4. Heatmap Organ.material (16s).png", width=7, height=6, units = "in", res=400)
htmp = psq %>%
	comp_heatmap(name = "Relative abundance", sample_names_show = T, numbers_use_counts = FALSE,
		numbers = heat_numbers(decimals = 3), cluster_columns = F, column_order = c("Branch", "Needles", "Shavings"),
		right_annotation = rowAnnotation(Enriched_in = diff_all, col = list(Enriched_in = col_fun), 
			show_annotation_name = c(Enriched_in = FALSE),
			annotation_legend_param = list(Enriched_in = list(title = "Enriched in"))),
		row_names_gp = gpar(fontsize = 8),
		row_names_max_width = max_text_width("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
		column_names_gp = gpar(fontsize = 9),
		column_names_rot = 0,
		column_names_centered = TRUE
	) 
draw(htmp, heatmap_legend_side = "bottom")
dev.off()

```
## Differential abundance test (Season)
```{r }
difab = clone(tmp)
tax = as.data.frame(difab$tax_table[,"Class_Genus"])
rownames(tax) = rownames(difab$tax_table)
colnames(tax) = "Class_Genus"
difab$tax_table = tax
difab = difab$merge_taxa("Class_Genus")

difab$tidy_dataset()
t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Season", taxa_level = "Class_Genus")

```

```{r }
diff = str_split(unique(filter(t1$res_diff, Factors == "SeasonWinter", passed_ss == T & diff == T)$Taxa), "\\|", simplify = TRUE)

psq <- meco2phyloseq(tmp)

psq = psq %>%
	merge_samples("Season") %>%
	tax_transform("compositional", rank = "Class_Genus") %>%
	tax_select(diff, ranks_searched = "Class_Genus", strict_matches = T) 

png("../fig S1. Heatmap Season (16s).png", width=7, height=6, units = "in", res=400)
htmp = psq %>%
	comp_heatmap(name = "Relative abundance", sample_names_show = T, numbers_use_counts = FALSE,
	numbers = heat_numbers(decimals = 3), cluster_columns = F, column_order = c("Summer", "Winter")) 
draw(htmp)
dev.off()

```

## Differential abundance test (Plant)
```{r }
difab = clone(tmp)
tax = as.data.frame(difab$tax_table[,"Class_Genus"])
rownames(tax) = rownames(difab$tax_table)
colnames(tax) = "Class_Genus"
difab$tax_table = tax
difab = difab$merge_taxa("Class_Genus")

difab$tidy_dataset()
t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Plant", taxa_level = "Class_Genus")

```

```{r }
diff = str_split(unique(filter(t1$res_diff, Factors == "PlantPj_2", passed_ss == T & diff == T)$Taxa), "\\|", simplify = TRUE)

psq <- meco2phyloseq(tmp)

psq = psq %>%
	merge_samples("Plant") %>%
	tax_transform("compositional", rank = "Class_Genus") %>%
	tax_select(diff, ranks_searched = "Class_Genus", strict_matches = T) 

png("Heatmap Plant (16s).png", width=7, height=6, units = "in", res=400)
htmp = psq %>%
	comp_heatmap(name = "Relative abundance", sample_names_show = T, numbers_use_counts = FALSE,
	numbers = heat_numbers(decimals = 3), cluster_columns = F, column_order = c("Pj_1", "Pj_2")) 
draw(htmp)
dev.off()

```

## Functional analysis
```{r}
load("plabase.RData")

ko <- read.delim("pred_metagenome_unstrat.tsv", row.names = 1)
ko = ko[,meco$sample_names()]

ko_rows = rownames(ko)
ko_rows = str_replace_all(ko_rows, "ko:", "")
ko = data.frame(lapply(ko, function(x) 
	if(is.numeric(x)) round(x) else x))
rownames(ko) = ko_rows

ko_pgpt = as_tibble(ko_rows) %>%
	left_join(plabase, by = "value") %>%
	mutate(level1 = ifelse(is.na(level1), "No PGPT", level1)) 

ko_pgpt = mutate(ko_pgpt, genes = ifelse(is.na(genes), "No PGPT", genes))
```

```{r }
ko_table = rowMeans(ko)
ko_table = as.data.frame(ko_table)
ko_table$value = rownames(ko_table)
colnames(ko_table) = c("abund", "value")

ko_table = left_join(ko_pgpt, ko_table, by = "value")

```
# PGPT and No PGPT genes
```{r }
ko_table %>%
	group_by(value, level1) %>%
	reframe() %>%
	mutate(abund = 1) %>%
	group_by(level1) %>%
	reframe(abund = sum(abund))
```
# Sankey diagramm
```{r }
df_top <- ko_table %>% 
	select(level1, level2, level3, level4, abund) %>%
	group_by(level1, level2, level3, level4) %>%
	reframe(abund = sum(abund)) %>%
	top_n(16, abund) 


df <- ko_table %>% 
	select(level1, level2, level3, level4, abund) %>%
	group_by(level1, level2, level3, level4) %>%
	reframe(abund = sum(abund)) %>%
	mutate(level2 = ifelse(level4 %in% df_top$level4, level2, paste0(level1, " OTHER"))) %>%
	mutate(level3 = ifelse(level2 %in% df_top$level2, level3, NA)) %>%
	mutate(level4 = ifelse(level2 %in% df_top$level2, level4, NA)) %>%
	group_by(level1, level2, level3, level4) %>%
	reframe(abund = sum(abund)) %>%
	ungroup() %>%
	filter(level1 != "No PGPT", level2 != "PUTATIVE_PGPTs") %>%
	# na.omit() %>%
	make_long(level1, level2, level3, level4, value = abund) %>%
	filter(node != "NA")
  
df$node <- factor(df$node,levels = rev(unique(df$node)))
df$next_node <- factor(df$next_node,levels = rev(unique(df$next_node)))

dagg <- df %>%
  group_by(node) %>%
  mutate(perc = value) %>%
  select(node, perc) %>%
  group_by(node) %>%
  reframe(perc = sum(perc)) %>%
  mutate(perc_PGPT = max(perc)) %>%
  mutate(perc = round(perc/perc_PGPT*100)) %>%
  select(node, perc) 
  


df2 <- merge(df, dagg, by.x = 'node', by.y = 'node', all.x = TRUE)

pl <- ggplot(df2, aes(x = x
                     , next_x = next_x
                     , node = node
                     , next_node = next_node
                     , fill = factor(node)
                     , label = paste0(node," (", perc, "%)")
					 , value = value)
             )
pl <- pl +geom_sankey(flow.alpha = 0.5
                      , node.color = "black"
                      ,show.legend = FALSE)
pl <- pl +geom_sankey_label(size = 5, color = "black", fill= "white", hjust = 0.5, vjust = -1)
pl <- pl +  theme_bw()
pl <- pl + theme(legend.position = "none")
pl <- pl +  theme(axis.title = element_blank()
                  , axis.text.y = element_blank()
				  , axis.text.x = element_text(size = 20)
                  , axis.ticks = element_blank()  
                  , panel.grid = element_blank())
pl <- pl + scale_fill_viridis_d(option = "inferno")
pl + 
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) 

ggsave("../fig 5. Functional analysis (16S).png", width=11, height=6, dpi=400, scale = 2)
```

## PGPT ontology level 5 microtable
```{r }
pathway_map = as.data.frame(ko_pgpt)


ko_path = ko
ko_path$value = rownames(ko_path)

pathway_abund = left_join(pathway_map, ko_path, by = "value") %>%
	pivot_longer(cols = rownames(meco$sample_table), names_to = "sample", values_to = "abund") %>%
	group_by(level5, sample) %>%
	reframe(abund = sum(abund)) %>%
	pivot_wider(names_from = "sample", values_from = "abund") %>%
	na.omit()

pathway_abund = as.data.frame(pathway_abund)
rownames(pathway_abund) = pathway_abund[,1]
pathway_abund = pathway_abund[,-1]

pathway = group_by(pathway_map, level5) %>%
		reframe() %>%
		na.omit()
pathway = as.data.frame(pathway)
rownames(pathway) = pathway[,1]

tmp <- microtable$new(otu_table = pathway_abund, tax_table = pathway, sample_table = meco$sample_table)
tmp$tidy_dataset()
tmp$filter_taxa(rel_abund = 0, freq = 0.20)
tmp
```

# Differential abundance test (PGPT ontology level 5)
```{r }
difab <- clone(tmp)
# difab$tax_table = filter(difab$tax_table, Class != "c__")
difab$tidy_dataset()
t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Organ.material", fix_formula = "Organ.material", taxa_level = "level5", pairwise = TRUE)
res_convert <- data.frame()
tmp_res = t1$res_diff_raw$res_pair
for(i in seq_len(nrow(tmp_res))){
	taxon_data <- tmp_res[i, ]
	raw_colnames <- colnames(taxon_data)
	all_factors <- raw_colnames %>% .[grepl("lfc_", .)] %>% gsub("lfc_", "", .)
	res_convert <- rbind(res_convert, data.frame(Taxa = tmp_res[i, "taxon"], 
		Factors = all_factors, 
		lfc = taxon_data %>% .[, grepl("^lfc_", raw_colnames)] %>% unlist,
		se = taxon_data %>% .[, grepl("^se_", raw_colnames)] %>% unlist,
		W = taxon_data %>% .[, grepl("^W_", raw_colnames)] %>% unlist,
		p = taxon_data %>% .[, grepl("^p_", raw_colnames)] %>% unlist,
		P.adj = taxon_data %>% .[, grepl("^q_", raw_colnames)] %>% unlist,
		diff = taxon_data %>% .[, grepl("^diff_", raw_colnames)] %>% unlist,
		passed_ss = taxon_data %>% .[, grepl("^passed_ss_", raw_colnames)] %>% unlist
		)
	)
}

t1$res_diff = res_convert

res = as_tibble(t1$res_diff) %>%
	arrange(Taxa) %>%
	group_by(Taxa) %>%
	filter(sum(P.adj < 0.05) >= 2) %>%
	ungroup()
```

```{r }
diff = unique(filter(res, passed_ss == T & diff == T)$Taxa)

Needles = unique(filter(res, Factors == "Organ.materialNeedles" & lfc > 0 & passed_ss == T & diff == T | 
	Factors == "Organ.materialShavings_Organ.materialNeedles" & lfc < 0 & passed_ss == T & diff == T )$Taxa )

Shavings = unique(filter(res, Factors == "Organ.materialShavings" & lfc > 0 & passed_ss == T & diff == T | 
	Factors == "Organ.materialShavings_Organ.materialNeedles" & lfc > 0 & passed_ss == T & diff == T )$Taxa)

Branch = unique(filter(res, Factors == "Organ.materialShavings" & lfc < 0 & passed_ss == T & diff == T  | 
	Factors == "Organ.materialNeedles" & lfc < 0 & passed_ss == T & diff == T )$Taxa )

Needles = as_tibble(Needles)
Needles$diff = "Needles"

Shavings = as_tibble(Shavings)
Shavings$diff = "Shavings"

Branch = as_tibble(Branch)
Branch$diff = "Branch"

all = rbind(Needles, Shavings, Branch) %>%
	arrange(value) %>%
	group_by(value) %>%
	nest(data = diff) %>%
	reframe(type = toString(unlist(data)))

diff_all = all$type
names(diff_all) = all$value

psq <- meco2phyloseq(tmp)

for (level in df_top$level4[!is.na(df_top$level4)]) {

select_path = filter(pathway_map, level4 == level) %>% 
	select(level5) %>% unique()
select_path = select_path$level5


psq_select = psq %>%
	merge_samples("Organ.material") %>%
	tax_transform("compositional", rank = "level5") %>%
	tax_select(select_path, ranks_searched = "level5", strict_matches = T) 

diff_all_select = diff_all[taxa_names(psq_select)]

col_fun = distinct_palette(n = length(unique(all$type)), add = NA)
names(col_fun) =  unique(all$type)

png(paste0("../", level, " func (16S).png"), width=7, height=6, units = "in", res=400)
htmp = psq_select %>%
	comp_heatmap(name = "Relative abundance", sample_names_show = T, numbers_use_counts = FALSE,
		numbers = heat_numbers(decimals = 3), cluster_columns = F, column_order = c("Branch", "Needles", "Shavings"),
		right_annotation = rowAnnotation(Enriched_in = diff_all_select, col = list(Enriched_in = col_fun), 
			show_annotation_name = c(Enriched_in = FALSE),
			annotation_legend_param = list(Enriched_in = list(title = "Enriched in"))),
		row_names_gp = gpar(fontsize = 8),
		row_names_max_width = max_text_width("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
		column_names_gp = gpar(fontsize = 9),
		column_names_rot = 0,
		column_names_centered = TRUE
	) 
draw(htmp, heatmap_legend_side = "bottom")
dev.off()
}
```
