```{r setup}
knitr::opts_chunk$set(
	error = TRUE,
	message = TRUE
)

mycolors = c("#018BFF", "#01FDFF", "#01FF61", "#F9FF01", "#FFD101", "#C08227", 
"#C02727", "#ecd078", "#C177C7", "#EBBAE3", "#BAEBC4", "#397DD7", "#6FE8D9", 
"#69d2e7", "#a7dbd8", "#e0e4cc", "#f38630", "#fa6900", "#fe4365", "#fc9d9a", "#f9cdad", 
"#c8c8a9", "#83af9b", "#65216B", "#d95b43", "#c02942", "#542437", "#53777a")
```

```{r}
set.seed(123)

library(phyloseq)
library(microeco)
library(file2meco)
library(tidyverse)
library(microbiome)
library(knitr)
library(DT)
library(RColorBrewer)
library(microViz)
library(ggsankey)
library(ComplexHeatmap)
library(ggpubr)
library(gridExtra)
library(grid)
library(forcats)
```
## Load phyloseq object
```{r}
physeq <- readRDS("physeq_ITS_187.rds")
asv = Biostrings::readDNAStringSet("asv.fasta")
physeq = merge_phyloseq(physeq, asv)
```
```{r}
metadata <- read.table("sampledata_ITS_meta.txt", header = T, row.names = 1, sep = "\t")
sample_data(physeq) <- metadata
stats_dada2 = readRDS("stats_dada2_ITS.rds")
metadata = merge(metadata, stats_dada2, by=0, all.x=TRUE)
rownames(metadata) = metadata[,1]
metadata$Tree_parts = factor(metadata$Tree_parts, c("Branch", "Needles", "Fresh_wood")) 
sample_data(physeq) <- metadata
```
## Transform phyloseq object to microtable class
```{r}
meco <- phyloseq2meco(physeq)
meco$sample_table = filter(meco$sample_table, Season != "Autumn")
meco$filter_pollution(taxa = c("mitochondria", "chloroplast"))
meco$tax_table = filter(meco$tax_table, Kingdom != "k__Eukaryota", Kingdom != "k__")
meco$tidy_dataset()
```
# Paired-end reads were obtained (sum)
```{r}
sum(meco$sample_table$input)
```
# Filtered sequances (sum)
```{r}
sum(meco$otu_table)
```
# Filtered sequances (mean)
```{r}
mean(colSums(meco$otu_table))
```
# Filtered sequances (median)
```{r}
median(colSums(meco$otu_table))
```

<!-- ```{r}
sampledata = merge(meco$sample_table, as.data.frame(colSums(meco$otu_table)), by=0, all.x=TRUE)
write.table(sampledata, "sampledata_ITS_mod.txt", quote=F, sep = "\t")
``` -->

## Fix tax table
```{r}
tax_meta = meco$tax_table

tax_rows = rownames(tax_meta)
t = apply(tax_meta, 2, function(x) str_replace_all(x, ".__$", "uncultured"))
rownames(t) = tax_rows
tax_meta = t

# Swap values ​​from Swap_unidentified in tax_meta to previous taxon level
Swap_unidentified = c("uncultured", "unidentified", "metagenome", "bacteriap25", "Unknown", "Incertae_Sedis")
for(Swap in Swap_unidentified) {
    for(j in 1:7) {
        for (i in 1:nrow(tax_meta)) {
            if(grepl(Swap, tax_meta[i,j])) {
                tax_meta[i,j] = tax_meta[i,j-1] }
                else { tax_meta[i,j] = tax_meta[i,j]
            }       
        }
    }
}
tax_meta = as.data.frame(tax_meta)

# Mark unclassified taxa at the genus level as "un."
tax_meta$Phylum = str_replace_all(tax_meta$Phylum, "k__.*", paste(tax_meta$Phylum, "un."))
tax_meta$Class = str_replace_all(tax_meta$Class, "k__.*", paste(tax_meta$Class, "un."))
tax_meta$Class = str_replace_all(tax_meta$Class, "p__.*", paste(tax_meta$Class, "un."))
tax_meta$Order = str_replace_all(tax_meta$Order, "k__.*", paste(tax_meta$Order, "un."))
tax_meta$Order = str_replace_all(tax_meta$Order, "p__.*", paste(tax_meta$Order, "un."))
tax_meta$Order = str_replace_all(tax_meta$Order, "c__.*", paste(tax_meta$Order, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "k__.*", paste(tax_meta$Family, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "p__.*", paste(tax_meta$Family, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "c__.*", paste(tax_meta$Family, "un."))
tax_meta$Family = str_replace_all(tax_meta$Family, "o__.*", paste(tax_meta$Family, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "k__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "p__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "c__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "o__.*", paste(tax_meta$Genus, "un."))
tax_meta$Genus = str_replace_all(tax_meta$Genus, "f__.*", paste(tax_meta$Genus, "un."))

tax_meta$Class_Genus = paste0(tax_meta$Class, "; ", tax_meta$Genus)
tax_meta = tax_meta[,c("Kingdom", "Phylum", "Class", "Order", "Family", "Class_Genus", "Genus", "Species")]

Species = paste("[ASV", c(1:length(rownames(meco$tax_table))), "]", sep = "")
Species = paste(Species, tax_meta$Genus, sep = " ")
meco$tax_table = as.data.frame(tax_meta)
meco$tax_table$Species = Species

```

## Rarefaction
```{r}
meco_rf <- clone(meco)
meco_rf$rarefy_samples(sample.size = 10000)
meco_rf$tidy_dataset()
```

# Alpha diversity
```{r}
tmp <- clone(meco_rf)
```
## Shannon (Tree_parts)
```{r fig.width=5, fig.height=6}
t1 <- trans_alpha$new(dataset = tmp, group = "Tree_parts")
t1$cal_diff(method = "anova")
t1$plot_alpha(measure = "Shannon") +
	scale_y_continuous(breaks = c(0,1,2,3,4,5,6), limits = c(2,6))
ggsave("../fig 7a. Shannon (Tree_parts) (ITS).png", width=3, height=4, dpi=400)
```
## Shannon (Season)
```{r fig.width=5, fig.height=6}
t1 <- trans_alpha$new(dataset = tmp, group = "Season")
t1$cal_diff(method = "anova")
t1$plot_alpha(measure = "Shannon") +
	scale_y_continuous(breaks = c(0,1,2,3,4,5,6), limits = c(2,6))
ggsave("../fig 7b. Shannon (Season) (ITS).png", width=3, height=4, dpi=400)
```
## Shannon (Tree_specimens)
```{r fig.width=5, fig.height=6}
t1 <- trans_alpha$new(dataset = tmp, group = "Tree_specimens")
t1$cal_diff(method = "anova")
t1$plot_alpha(measure = "Shannon") +
	scale_y_continuous(breaks = c(0,1,2,3,4,5,6), limits = c(2,6))
ggsave("../fig 7c. Shannon (Tree_specimens) (ITS).png", width=3, height=4, dpi=400)
```

# Beta diversity
```{r }
tmp <- clone(meco_rf)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
tmp$tidy_dataset()
tmp$cal_betadiv()
```
## Bray-curtis (Tree_parts)
```{r fig.width=6, fig.height=6}
t1 <- trans_beta$new(dataset = tmp, group = "Tree_parts", measure = "bray")
t1$cal_ordination(method = "PCoA")
t1$plot_ordination(plot_color = "Tree_parts", plot_shape = "Tree_specimens", plot_type = c("point", "ellipse")) + 
	guides(
		color = guide_legend(order = 1),
		fill = guide_legend(order = 1),
		shape = guide_legend(order = 2)
	) + 
	labs(
		color = "Tree_parts",
		fill = "Tree_parts"
	)
ggsave("../fig 7d. Bray-curtis (Tree_parts) (ITS).png", width=4, height=4, dpi=400)
```
## PERMANOVA (Tree_parts)
```{r }
t1$cal_manova(group = "Tree_parts", manova_all = F)
kable(t1$res_manova)
t1$cal_manova(group = "Tree_parts", manova_all = T)
kable(t1$res_manova)
```
## Bray-curtis (Season)
```{r fig.width=6, fig.height=6}
t1 <- trans_beta$new(dataset = tmp, group = "Season", measure = "bray")
t1$cal_ordination(method = "PCoA")
t1$plot_ordination(plot_color = "Season", plot_shape = "Tree_specimens", plot_type = c("point", "ellipse")) + 
	guides(
		color = guide_legend(order = 1),
		fill = guide_legend(order = 1),
		shape = guide_legend(order = 2)
	) + 
	labs(
		color = "Season",
		fill = "Season"
	)
ggsave("../fig 7e. Bray-curtis (Season) (ITS).png", width=4, height=4, dpi=400)
```
## PERMANOVA (Season)
```{r }
t1$cal_manova(group = "Season", manova_all = F)
kable(t1$res_manova)
```
## Bray-curtis (Tree_specimens)
```{r fig.width=6, fig.height=6}
t1 <- trans_beta$new(dataset = tmp, group = "Tree_specimens", measure = "bray")
t1$cal_ordination(method = "PCoA")
t1$plot_ordination(plot_color = "Tree_specimens", plot_shape = "Tree_parts", plot_type = c("point", "ellipse")) + 
	guides(
		color = guide_legend(order = 1),
		fill = guide_legend(order = 1),
		shape = guide_legend(order = 2)
	) + 
	labs(
		color = "Tree_specimens",
		fill = "Tree_specimens"
	)
ggsave("../fig 7f. Bray-curtis (Tree_specimens) (ITS).png", width=4, height=4, dpi=400)
```
## PERMANOVA (Tree_specimens)
```{r }
t1$cal_manova(group = "Tree_specimens", manova_all = F)
kable(t1$res_manova)
```

# ASV table
```{r}
tmp <- clone(meco_rf)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
ASV = as.data.frame(tmp$rep_fasta)
ASV$ASV = rownames(ASV)
Taxonomy = tmp$tax_table
Taxonomy$ASV = rownames(Taxonomy)
ASV = left_join(ASV, Taxonomy, by = "ASV")
ASV = as_tibble(ASV)

bar = clone(tmp)
bar = bar$merge_samples("Tree_parts")
t1 <- trans_abund$new(dataset = bar, taxrank = "Species", ntaxa = 1000, delete_taxonomy_prefix = F)

Tree_parts = pivot_wider(t1$data_abund[,c("Taxonomy", "Sample", "Abundance")], names_from = "Sample", values_from = "Abundance")
colnames(Tree_parts) = c("Species", "Needles", "Fresh_wood", "Branch")

bar = clone(tmp)
bar = bar$merge_samples("Season")
t1 <- trans_abund$new(dataset = bar, taxrank = "Species", ntaxa = 1000, delete_taxonomy_prefix = F)

Season = pivot_wider(t1$data_abund[,c("Taxonomy", "Sample", "Abundance")], names_from = "Sample", values_from = "Abundance")
colnames(Season) = c("Species", "Summer", "Winter")

bar = clone(tmp)
bar = bar$merge_samples("Tree_specimens")
t1 <- trans_abund$new(dataset = bar, taxrank = "Species", ntaxa = 1000, delete_taxonomy_prefix = F)

Tree_specimens = pivot_wider(t1$data_abund[,c("Taxonomy", "Sample", "Abundance")], names_from = "Sample", values_from = "Abundance")
colnames(Tree_specimens) = c("Species", "Pj_1", "Pj_2")

ASV = left_join(ASV, Tree_parts, by = "Species")
ASV = left_join(ASV, Season, by = "Species")
ASV = left_join(ASV, Tree_specimens, by = "Species")
ASV = ASV[, c("x", "Kingdom","Phylum","Class","Order","Family","Genus","Species","Needles","Fresh_wood","Branch","Summer","Winter","Pj_1","Pj_2")]
colnames(ASV) = c("DNA sequence", "Kingdom","Phylum","Class","Order","Family","Genus","ASV","Needles","Fresh wood","Branch","Summer","Winter","Pj_1","Pj_2")

ASV[, !names(ASV) %in% "ASV"] = as_tibble(apply(ASV[, !names(ASV) %in% "ASV"], 2, function(x) str_replace_all(x, ".*un\\.", "__")))
ASV$ASV = str_replace_all(ASV$ASV, " un.", "")

write.table(ASV, "../Table S9. ASVs (ITS).txt", quote = F, sep = "\t")
```

# Composition
```{r}
tmp <- clone(meco_rf)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
tmp$tax_table$Class = str_replace_all(tmp$tax_table$Class, "k__.*", "k__")
tmp$tax_table$Class = str_replace_all(tmp$tax_table$Class, "p__.*", "p__")
tmp$tax_table$Family = str_replace_all(tmp$tax_table$Family, "k__.*", "k__")
tmp$tax_table$Family = str_replace_all(tmp$tax_table$Family, "p__.*", "p__")
tmp$tax_table$Family = str_replace_all(tmp$tax_table$Family, "c__.*", "c__")
tmp$tax_table$Family = str_replace_all(tmp$tax_table$Family, "o__.*", "o__")
tmp$tax_table$Order = str_replace_all(tmp$tax_table$Order, "k__.*", "k__")
tmp$tax_table$Order = str_replace_all(tmp$tax_table$Order, "p__.*", "p__")
tmp$tax_table$Order = str_replace_all(tmp$tax_table$Order, "c__.*", "c__")
tmp$tidy_dataset()
```



## Tree_parts bar plot (Class)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Tree_parts")
t1 <- trans_abund$new(dataset = bar, taxrank = "Class", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
barplot <- function(order_x, order_tax, path) {

p = t1$plot_bar(others_color = "grey70", xtext_keep = TRUE, legend_text_italic = FALSE,
	color_values = mycolors,
	order_x = order_x, xtext_angle = 45) +
geom_text(aes(x = Sample, y = Abundance, 
		label = ifelse(Abundance > 5, scales::percent(Abundance, scale = 1, accuracy = 1), ""), 
		group = Taxonomy),
		inherit.aes = FALSE, position = position_stack(vjust = 0.5), col = "black") +
scale_fill_manual(
    values = c(mycolors[1:length(order_tax)-1], "grey70"),
    breaks = order_tax) +
guides(fill = guide_legend(reverse = TRUE))

legend <- get_legend(p)
plot_without_legend <- p + theme(legend.position = "none")

ggarrange(plot_without_legend)
ggsave(paste0(path, ".png"), width=2, height=4, dpi=400)

ggarrange(legend)
ggsave(paste0(path, " legend.png"), width=2, height=4, dpi=400)

}

path = "../fig 8a. Tree_parts bar plot (Class) (ITS)"
order_x = c("Branch", "Needles", "Fresh_wood")
order_tax = c("c__Dothideomycetes", "c__Sordariomycetes", "c__Eurotiomycetes", 
	"c__Leotiomycetes", "c__Malasseziomycetes", "c__Lecanoromycetes",
	"c__Tremellomycetes", "c__Agaricomycetes", "Others")
barplot(order_x, order_tax, path)
```

## Tree_parts bar plot (Order)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Tree_parts")
t1 <- trans_abund$new(dataset = bar, taxrank = "Order", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8d. Tree_parts bar plot (Order) (ITS)"
order_x = c("Branch", "Needles", "Fresh_wood")
order_tax = c("o__Mycosphaerellales", "o__Pleosporales", "o__Venturiales", 
	"o__Xylariales", "o__Capnodiales", "o__Chaetothyriales",
	"o__Helotiales", "o__Eurotiales", "o__Malasseziales",
	"o__Mytilinidiales", "Others")
barplot(order_x, order_tax, path)
```

## Tree_parts bar plot (Family)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Tree_parts")
t1 <- trans_abund$new(dataset = bar, taxrank = "Family", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8g. Tree_parts bar plot (Family) (ITS)"
order_x = c("Branch", "Needles", "Fresh_wood")
order_tax = c("f__Teratosphaeriaceae", "f__Venturiaceae", "f__Cylindrosympodiaceae",
	"f__Lentitheciaceae", "f__Trichomeriaceae", "f__Anteagloniaceae",
	"f__Aspergillaceae", "f__Malasseziaceae", "f__Mytilinidiaceae",
	"f__Teloschistaceae", "Others")
barplot(order_x, order_tax, path)
```

## Tree_parts venn diagramm
```{r fig.width=5, fig.height=5}
ven = clone(bar)

tax = as.data.frame(ven$tax_table[,"Species"])
rownames(tax) = rownames(ven$tax_table)
colnames(tax) = "Species"
ven$tax_table = tax
ven = ven$merge_taxa("Species")

rownames(ven$tax_table) = ven$tax_table$Species
rownames(ven$otu_table) = ven$tax_table$Species
t1 <- trans_venn$new(ven, ratio = "seqratio")
t1$plot_venn()
ggsave("../fig 9a. Tree_parts venn diagramm (ASV) (ITS).png", width=5, height=5, dpi=400)
```
```{r}
datatable(t1$data_details, options = list(scrollX = TRUE, scrollY = TRUE))
write.table(t1$data_details, sep = "\t", quote = FALSE, file="Tree_parts venn diagramm (ASV) (ITS).txt")
```

## Season bar plot (Class)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Season")
t1 <- trans_abund$new(dataset = bar, taxrank = "Class", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8b. Season bar plot (Class) (ITS)"
order_x = c("Summer", "Winter")
order_tax = c("c__Dothideomycetes", "c__Sordariomycetes", "c__Eurotiomycetes", 
	"c__Leotiomycetes", "c__Malasseziomycetes", "c__Lecanoromycetes",
	"c__Tremellomycetes", "c__Agaricomycetes", "Others")
barplot(order_x, order_tax, path)
```

## Season bar plot (Order)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Season")
t1 <- trans_abund$new(dataset = bar, taxrank = "Order", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8e. Season bar plot (Order) (ITS)"
order_x = c("Summer", "Winter")
order_tax = c("o__Mycosphaerellales", "o__Pleosporales", "o__Venturiales", 
	"o__Xylariales", "o__Capnodiales", "o__Chaetothyriales",
	"o__Helotiales", "o__Eurotiales", "o__Malasseziales",
	"o__Mytilinidiales", "Others")
barplot(order_x, order_tax, path)
```

## Season bar plot (Family)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Season")
t1 <- trans_abund$new(dataset = bar, taxrank = "Family", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8h. Season bar plot (Family) (ITS)"
order_x = c("Summer", "Winter")
order_tax = c("f__Teratosphaeriaceae", "f__Venturiaceae", "f__Cylindrosympodiaceae",
	"f__Lentitheciaceae", "f__Trichomeriaceae", "f__Anteagloniaceae",
	"f__Aspergillaceae", "f__Malasseziaceae", "f__Mytilinidiaceae",
	"f__Teloschistaceae", "Others")
barplot(order_x, order_tax, path)
```

## Season venn diagramm
```{r fig.width=5, fig.height=5}
ven = clone(bar)

tax = as.data.frame(ven$tax_table[,"Species"])
rownames(tax) = rownames(ven$tax_table)
colnames(tax) = "Species"
ven$tax_table = tax
ven = ven$merge_taxa("Species")

rownames(ven$tax_table) = ven$tax_table$Species
rownames(ven$otu_table) = ven$tax_table$Species
t1 <- trans_venn$new(ven, ratio = "seqratio")
t1$plot_venn()
ggsave("../fig 9b. Season venn diagramm (ASV) (ITS).png", width=5, height=5, dpi=400)
```
```{r}
datatable(t1$data_details, options = list(scrollX = TRUE, scrollY = TRUE))
write.table(t1$data_details, sep = "\t", quote = FALSE, file="Season venn diagramm (ASV) (ITS).txt")
```

## Tree_specimens bar plot (Class)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Tree_specimens")
t1 <- trans_abund$new(dataset = bar, taxrank = "Class", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8c. Tree_specimens bar plot (Class) (ITS)"
order_x = c("Pj_1", "Pj_2")
order_tax = c("c__Dothideomycetes", "c__Sordariomycetes", "c__Eurotiomycetes", 
	"c__Leotiomycetes", "c__Malasseziomycetes", "c__Lecanoromycetes",
	"c__Tremellomycetes", "c__Agaricomycetes", "Others")
barplot(order_x, order_tax, path)
```

## Tree_specimens bar plot (Order)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Tree_specimens")
t1 <- trans_abund$new(dataset = bar, taxrank = "Order", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8f. Tree_specimens bar plot (Order) (ITS)"
order_x = c("Pj_1", "Pj_2")
order_tax = c("o__Mycosphaerellales", "o__Pleosporales", "o__Venturiales", 
	"o__Xylariales", "o__Capnodiales", "o__Chaetothyriales",
	"o__Helotiales", "o__Eurotiales", "o__Malasseziales",
	"o__Mytilinidiales", "Others")
barplot(order_x, order_tax, path)
```

## Tree_specimens bar plot (Family)
```{r}
bar = clone(tmp)
bar = bar$merge_samples("Tree_specimens")
t1 <- trans_abund$new(dataset = bar, taxrank = "Family", ntaxa = 10, delete_taxonomy_prefix = F)
```
```{r fig.width=5, fig.height=6}
path = "../fig 8i. Tree_specimens bar plot (Family) (ITS)"
order_x = c("Pj_1", "Pj_2")
order_tax = c("f__Teratosphaeriaceae", "f__Venturiaceae", "f__Cylindrosympodiaceae",
	"f__Lentitheciaceae", "f__Trichomeriaceae", "f__Anteagloniaceae",
	"f__Aspergillaceae", "f__Malasseziaceae", "f__Mytilinidiaceae",
	"f__Teloschistaceae", "Others")
barplot(order_x, order_tax, path)
```

## Tree_specimens venn diagramm
```{r fig.width=5, fig.height=5}
ven = clone(bar)

tax = as.data.frame(ven$tax_table[,"Species"])
rownames(tax) = rownames(ven$tax_table)
colnames(tax) = "Species"
ven$tax_table = tax
ven = ven$merge_taxa("Species")

rownames(ven$tax_table) = ven$tax_table$Species
rownames(ven$otu_table) = ven$tax_table$Species
t1 <- trans_venn$new(ven, ratio = "seqratio")
t1$plot_venn()
ggsave("../fig 9c. Tree_specimens venn diagramm (ASV) (ITS).png", width=5, height=5, dpi=400)
```
```{r}
datatable(t1$data_details, options = list(scrollX = TRUE, scrollY = TRUE))
write.table(t1$data_details, sep = "\t", quote = FALSE, file="Tree_specimens venn diagramm (ASV) (ITS).txt")
```

## Core genera
```{r}
core = clone(tmp)

tax = as.data.frame(core$tax_table[,"Class_Genus"])
rownames(tax) = rownames(core$tax_table)
colnames(tax) = "Class_Genus"
core$tax_table = tax
core = core$merge_taxa("Class_Genus")

core <- meco2phyloseq(core)
pseq.rel <- microbiome::transform(core, "compositional")

taxa_names(pseq.rel) = tax_table(pseq.rel)[,"Class_Genus"]

pseq.rel = subset_taxa(pseq.rel, !(Class_Genus %in% str_extract(tax_table(pseq.rel)[,"Class_Genus"], ".*un.")))

# Core with compositionals:
prevalences <- seq(.00, 1, .1)
# detections <- round(10^seq(log10(1e-2), log10(.2), length = 20), 3)
detections <- c(0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.08, 0.10, 0.12, 0.14, 0.16)

# Also define gray color palette
gray <- gray(seq(0,1,length=5))

p1 <- plot_core(pseq.rel,
  plot.type = "heatmap",
  colours = rev(brewer.pal(5, "RdBu")),
  prevalences = prevalences,
  detections = detections, min.prevalence = .5) +
  xlab("Detection Threshold (Relative Abundance (%))")

p1 <- p1 + theme_bw() + ylab("") + scale_x_discrete(labels = round(detections * 100, 0)) +
	scale_fill_gradientn(labels = scales::percent, breaks = prevalences, colours = rev(brewer.pal(5, "RdBu")),
    values = c(0, 0.25, 0.5, 0.75, 1)) + labs(fill = "Frequency of\noccurrence")
p1
ggsave("../fig 10. core (ITS).png", width=8, height=3, dpi=400)
```

## Differential abundance test (Tree_parts)
```{r}
tmp <- clone(meco)
tmp$filter_taxa(rel_abund = 0.001, freq = 0.2)
tmp$tidy_dataset()
```
```{r }
difab = clone(tmp)
tax = as.data.frame(difab$tax_table[,"Class_Genus"])
rownames(tax) = rownames(difab$tax_table)
colnames(tax) = "Class_Genus"
difab$tax_table = tax
difab = difab$merge_taxa("Class_Genus")

t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Tree_parts", fix_formula = "Tree_parts", taxa_level = "Class_Genus", pairwise = TRUE)

res_convert <- data.frame()
tmp_res = t1$res_diff_raw$res_pair
for(i in seq_len(nrow(tmp_res))){
	taxon_data <- tmp_res[i, ]
	raw_colnames <- colnames(taxon_data)
	all_factors <- raw_colnames %>% .[grepl("lfc_", .)] %>% gsub("lfc_", "", .)
	res_convert <- rbind(res_convert, data.frame(Taxa = tmp_res[i, "taxon"], 
		Factors = all_factors, 
		lfc = taxon_data %>% .[, grepl("^lfc_", raw_colnames)] %>% unlist,
		se = taxon_data %>% .[, grepl("^se_", raw_colnames)] %>% unlist,
		W = taxon_data %>% .[, grepl("^W_", raw_colnames)] %>% unlist,
		p = taxon_data %>% .[, grepl("^p_", raw_colnames)] %>% unlist,
		P.adj = taxon_data %>% .[, grepl("^q_", raw_colnames)] %>% unlist,
		diff = taxon_data %>% .[, grepl("^diff_", raw_colnames)] %>% unlist,
		passed_ss = taxon_data %>% .[, grepl("^passed_ss_", raw_colnames)] %>% unlist
		)
	)
}

t1$res_diff = res_convert

res = as_tibble(t1$res_diff) %>%
	arrange(Taxa) %>%
	group_by(Taxa) %>%
	filter(sum(P.adj < 0.05) >= 2) %>%
	ungroup()
```

```{r }
diff = str_split(unique(filter(res, passed_ss == T & diff == T)$Taxa), "\\|", simplify = TRUE)

Needles = str_split(unique(filter(res, Factors == "Tree_partsNeedles" & lfc > 0 & passed_ss == T & diff == T | 
	Factors == "Tree_partsFresh_wood_Tree_partsNeedles" & lfc < 0 & passed_ss == T & diff == T )$Taxa ), "\\|", simplify = TRUE)

Fresh_wood = str_split(unique(filter(res, Factors == "Tree_partsFresh_wood" & lfc > 0 & passed_ss == T & diff == T | 
	Factors == "Tree_partsFresh_wood_Tree_partsNeedles" & lfc > 0 & passed_ss == T & diff == T )$Taxa ), "\\|", simplify = TRUE)

Branch = str_split(unique(filter(res, Factors == "Tree_partsFresh_wood" & lfc < 0 & passed_ss == T & diff == T  | 
	Factors == "Tree_partsNeedles" & lfc < 0 & passed_ss == T & diff == T )$Taxa ), "\\|", simplify = TRUE)

Needles = as_tibble(Needles)
Needles$diff = "Needles"

Fresh_wood = as_tibble(Fresh_wood)
Fresh_wood$diff = "Fresh_wood"

Branch = as_tibble(Branch)
Branch$diff = "Branch"

all = rbind(Needles, Fresh_wood, Branch) %>%
	arrange(V1) %>%
	group_by(V1) %>%
	nest(data = diff) %>%
	reframe(type = toString(unlist(data)))

diff_all = all$type
names(diff_all) = all$V1

psq <- meco2phyloseq(tmp)

psq = psq %>%
	merge_samples("Tree_parts") %>%
	tax_transform("compositional", rank = "Class_Genus") %>%
	tax_select(diff, ranks_searched = "Class_Genus", strict_matches = T) 

psq = subset_taxa(psq, !(Class_Genus %in% str_extract(tax_table(psq)[,"Class_Genus"], ".*un.")))

diff_all = diff_all[taxa_names(psq)]

col_fun = distinct_palette(n = length(unique(all$type)), add = NA)
names(col_fun) =  unique(all$type)

png("../fig 11. Heatmap Tree_parts (ITS).png", width=7, height=6, units = "in", res=400)
htmp = psq %>%
	comp_heatmap(name = "Relative abundance", sample_names_show = T, numbers_use_counts = FALSE,
		numbers = heat_numbers(decimals = 3), cluster_columns = F, column_order = c("Branch", "Needles", "Fresh_wood"),
		right_annotation = rowAnnotation(Enriched_in = diff_all, col = list(Enriched_in = col_fun), 
			show_annotation_name = c(Enriched_in = FALSE),
			annotation_legend_param = list(Enriched_in = list(title = "Enriched in"))),
		row_names_gp = gpar(fontsize = 8),
		row_names_max_width = max_text_width("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
		column_names_gp = gpar(fontsize = 9),
		column_names_rot = 0,
		column_names_centered = TRUE
	) 
draw(htmp, heatmap_legend_side = "bottom")
dev.off()
```
## Differential abundance test (Season)
```{r }
difab = clone(tmp)
tax = as.data.frame(difab$tax_table[,"Class_Genus"])
rownames(tax) = rownames(difab$tax_table)
colnames(tax) = "Class_Genus"
difab$tax_table = tax
difab = difab$merge_taxa("Class_Genus")

difab$tidy_dataset()
t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Season", taxa_level = "Class_Genus")

```

```{r }
diff = str_split(unique(filter(t1$res_diff, Factors == "SeasonWinter", passed_ss == T & diff == T)$Taxa), "\\|", simplify = TRUE)
diff
```

## Differential abundance test (Tree_specimens)
```{r }
difab = clone(tmp)
tax = as.data.frame(difab$tax_table[,"Class_Genus"])
rownames(tax) = rownames(difab$tax_table)
colnames(tax) = "Class_Genus"
difab$tax_table = tax
difab = difab$merge_taxa("Class_Genus")

difab$tidy_dataset()
t1 <- trans_diff$new(dataset = difab, method = "ancombc2", group = "Tree_specimens", taxa_level = "Class_Genus")

```

```{r }
diff = str_split(unique(filter(t1$res_diff, Factors == "Tree_specimensPj_2", passed_ss == T & diff == T)$Taxa), "\\|", simplify = TRUE)

psq <- meco2phyloseq(tmp)

psq = psq %>%
	merge_samples("Tree_specimens") %>%
	tax_transform("compositional", rank = "Class_Genus") %>%
	tax_select(diff, ranks_searched = "Class_Genus", strict_matches = T) 

psq = subset_taxa(psq, !(Class_Genus %in% str_extract(tax_table(psq)[,"Class_Genus"], ".*un.")))

png("../fig S12. Heatmap Tree_specimens (ITS).png", width=7, height=6, units = "in", res=400)
htmp = psq %>%
	comp_heatmap(name = "Relative abundance", sample_names_show = T, numbers_use_counts = FALSE,
	numbers = heat_numbers(decimals = 3), cluster_columns = F, column_order = c("Pj_1", "Pj_2")) 
draw(htmp)
dev.off()

```
